ENH: enable `H5NetCDFStore` to work with already open h5netcdf.File a…

kmuehlbauer opened the pull request:
enable `H5NetCDFStore` to work with open h5netcdf.File and h5netcdf.Group objects, add test

<!-- Feel free to remove check-list items aren't relevant to your change -->

 - [x] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

kmuehlbauer commented:
I'm unsure how this should be referenced in the docs/news because the netCDF4 counterpart doesn't have any mention about that feature too.

It seems that everything runs smoothly, the one error looks like it is unrelated. Please let me know, if the added test is enough and if you want me to do anything else. I'll leave the `WIP` mark in place so far.

Ping @shoyer, he might see if there are immediate problems with the code.

kmuehlbauer commented:
Happy New Year to everyone especially at pydata.

I'd be happy to actively follow any reviews to sort things out and to get this addition pulled in. Many thanks!

shoyer commented on xarray/backends/h5netcdf_.py:
```diff
@@ -69,8 +69,40 @@ class H5NetCDFStore(WritableCFDataStore):
+
+        if isinstance(manager, (h5netcdf.File, h5netcdf.Group)):
+            if group is None:
+                root, group = manager._root, manager.name
```
I think it's probably a better idea to use the `find_root_and_group` helper function that we use for netCDF4. Accessing private attributes like `_root` is a little unsafe.

kmuehlbauer replied:
#3680 needs to be resolved first. I appreciated any guidance on that blocking issue.

shoyer commented on xarray/backends/h5netcdf_.py:
```diff
@@ -69,8 +69,40 @@ class H5NetCDFStore(WritableCFDataStore):
+        self._group = group
+        self._mode = mode
+        self.format = None
+        self._filename = self.ds._root.filename
```
Likewise, we don't want to rely on a private API here

kmuehlbauer replied:
Unfortunately `h5netcdf.Group` doesn't seem to have a `filename` property like `netCDF4.Dataset`. So there is currently no direct means of extracting the filename from `h5netcdf.Group` using public API. Trying to to facilitate `find_root_and_group` now.

kmuehlbauer commented:
@TomNicholas After commit #1689db493f10262555196f658c52e370aacb4a33 I get mypy TypCheck error:

xarray/core/dataset.py:3610: error: Item "Mapping[...]" of "Union[Dataset, Mapping[...]]" has no attribute "to_dataset"

Should I create an issue on that?

kmuehlbauer commented:
@shoyer I tried to address your review comments. I've added a temporary (backwards compatible) fix for #3680 to get along. Also using `find_root_and_group` to get hold of the filename is some kind of workaround until there is a better option.

I've added an item to `whats-new.rst`. For the API docs I did not find a fitting position because `H5NetCDFStore` and `NetCDF4DataStore` are nowhere mentioned there. So this might better be added in another PR by experienced xarray devs.

Unfortunately mypy checks are breaking due to some recent change (PR #3677).

kmuehlbauer commented:
@shoyer Please let me know if this needs rebasing after #3690 gets merged, or if I can anything do to finalize this.

kmuehlbauer commented:
Took the chance to rebase on latest master.

kmuehlbauer commented:
@dcherian Following up on #3702 it would be really great if this enhancement could make it into 0.15.0 If I can be of any further assistance please ping me.

shoyer commented:
This looks great, thanks for figuring this out!
